girder/cli/mount.py

```diff
@@ -112,18 +117,30 @@ def _getPath(self, path):
         # '/user', and 'collection' before calling this method.
         if '/' not in path.rstrip('/')[1:]:
             raise fuse.FuseOSError(errno.ENOENT)
-        try:
-            # We can't filter the resource, since that removes files'
-            # assetstore information and users' size information.
-            resource = path_util.lookUpPath(
-                path.rstrip('/'), filter=False, force=True)
-        except (path_util.NotFoundException, AccessException):
-            raise fuse.FuseOSError(errno.ENOENT)
-        except ValidationException:
-            raise fuse.FuseOSError(errno.EROFS)
-        except Exception:
-            logger.exception('ServerFuse server internal error')
-            raise fuse.FuseOSError(errno.EROFS)
+        cacheEntry = self._getPathCache.get(path)
+        if cacheEntry and cacheEntry['expiry'] > time.time():
+            resource = cacheEntry['result']
+            cacheEntry['hits'] += 1
+        else:
+            try:
+                # We can't filter the resource, since that removes files'
+                # assetstore information and users' size information.
+                resource = path_util.lookUpPath(
+                    path.rstrip('/'), filter=False, force=True)
+                if len(self._getPathCache) > self._getPathCacheMaxEntries:
+                    self._getPathCache.clear()
```

Comment:
LRU behavior is probably preferable, but would require an additional data structure, so might not be worth the complexity.

Is there a reason not to make use of the cache utils present in Girder core?

Reply:
Sure, LRU with TTL would be nicer, but this is simple and sufficient. 

I didn't use the dogpile Memory backend because it seemed like it was more complex, and, also, I don't think it would have a virtue to allow it to use a non-memory backend.

Reply:
This is fine, as long as a sawtooth performance pattern is acceptable for users.

My one concern would be whether this behavior is sensible as a default, or whether this is something that should be configurable (e.g. env vars).

Do you have any measurements of the performance gain for your use cases?

Reply:
> I don't think it would have a virtue to allow it to use a non-memory backend.

Like what?

Reply:
It saves 10-11ms per served tile in one test using large_image.

Reply:
Looking into using dogpile a bit further: 

One could add:
```
from dogpile.cache import make_region

_getPathCache = make_region().configure(backend='dogpile.cache.memory', expiration_time=1)

class ServerFuse ...

    @_getPathCache.cache_on_arguments()
    def _getPath(self, path):
        ...
```
which works.  But, dogpile's memory backend has [caveats](https://dogpilecache.sqlalchemy.org/en/latest/api.html):

"There is no size management, and values which are placed into the dictionary will remain until explicitly removed. Note that Dogpile’s expiration of items is based on timestamps and does not remove them from the cache."

which seem like a big enough issue not to use it out-of-the-box.

Reply:
I guess it's too much to ask for a caching library to implement basic eviction policies in one of its core backends...

Reply:
Others have discussed using cachetools with dogpile ([see here](https://bitbucket.org/zzzeek/dogpile.cache/pull-requests/32/add-a-cachetools-lru-lfu-in-memory-backend/diff)), but that went nowhere.  Since then dogpile.cache has been moved to sqlalchemy's organization and nothing changed.

Reply:
I'm concerned that we're starting on the path to implementing and being forced to maintain a new Girder-specific code stack for what should be delegated to a mature upstream library. If the "cache" was just a global dict, then perhaps it wouldn't be worthwhile to pull in heavier dependencies, but even this implementation has eviction policies, size limits, etc. I suspect that we'll want even more sophistication and usage of caching in the future, leading to a situation where Girder's Python client and Python server have totally divergent caching technology.

If a well-established upstream library is lacking a necessary feature, I don't think the first-line solution should be to re-implement the entire system from scratch. We should probably implement a new Dogpile backend with the necessary feature(s), and perhaps publish it in a separate package / repo, so all of our projects (and even other external users) can benefit.

Reply:
As always, there is a balance -- There are greater costs in developing things for other libraries.

How does dogpile's backend managing multi-threading, for instance?  Here, I am using a simple dictionary with specific access methods and can see at a glance that it will be robust in a multi-threaded environment without locks (assuming cpython, of course).  I know you can use a lock with dogpile, but acquiring and releasing a lock has a time cost, too.

We added dogpile specifically to try to offload caching behavior to an upstream library, but I don't observe that anything has actually used it since.  Were we premature in selecting it?  Should we have gone with plain cachetools instead?  If so, that requires using locks for thread safety which then involves local code anyway.

If we are trying to reduce maintenance burden, adding another repo is more cognitive burden.  This is worth it if it gets used multiple places but not if it ends up only being used by one spot in one project.

Reply:
We shouldn't need to use locks for an in-memory cache implementation with cpython; even if we need to implement synchronized data structures, I think doing so in a way that we are confident does not release the GIL would be fairly straightforward. I tend to agree that the desire for caching utilities is general enough that we should create / leverage a more central solution instead of ad hoc versions all over the place (especially since we already kind of have an API for this in core). I would find it hard to believe if we had to implement LRU behavior ourselves rather than using a third party library, but if that's the case I'd have fun creating that abstraction.